src/main/java/com/openlattice/chronicle/services/message/MessageService.kt

```diff
@@ -0,0 +1,118 @@
+import com.openlattice.chronicle.services.entitysets.EntitySetIdsManager
+import com.openlattice.chronicle.services.twilio.TwilioManager
+import com.openlattice.chronicle.util.ChronicleServerUtil
+import com.openlattice.client.ApiClient
+import com.openlattice.data.DataAssociation
+import com.openlattice.data.DataGraph
+import org.slf4j.LoggerFactory
+import java.time.OffsetDateTime
+import java.util.*
+import java.util.concurrent.ExecutionException
+
+/**
+ * @author toddbergman &lt;todd@openlattice.com&gt;
+ */
+open class MessageService(
+    private val apiCacheManager: ApiCacheManager,
+    private val edmCacheManager: EdmCacheManager,
+    private val enrollmentManager: EnrollmentManager,
+    private val entitySetIdsManager: EntitySetIdsManager,
+    private val twilioManager: TwilioManager
+) : MessageManager {
+
+    companion object {
+        protected val logger = LoggerFactory.getLogger(MessageService::class.java)
+    }
+
+    override fun sendMessages(organizationId: UUID, messageDetailsList: List<MessageDetails>) {
+        val outcomes = twilioManager.sendMessages(messageDetailsList)
+        recordMessagesSent(organizationId, outcomes)
+    }
+
+    @Throws(ExecutionException::class)
+    fun recordMessagesSent(organizationId: UUID?, messageOutcomes: List<MessageOutcome>) {
+        val apiClient = apiCacheManager.prodApiClientCache[ApiClient::class.java]
+        val dataApi = apiClient.dataApi
+        val entities: ListMultimap<UUID, Map<UUID, Set<Any>>> = ArrayListMultimap.create();
+        val associations: ListMultimap<UUID, DataAssociation> = ArrayListMultimap.create()
+        // entity set ids
+        messageOutcomes.forEachIndexed { index, messageOutcome ->
+            val participantES = ChronicleServerUtil.getParticipantEntitySetName(messageOutcome.studyId)
+            val coreAppConfig = entitySetIdsManager
+                .getChronicleAppConfig(organizationId)
+            val legacyAppConfig = entitySetIdsManager
+                .getLegacyChronicleAppConfig(participantES)
+
+            val messageESID = coreAppConfig.messagesEntitySetId
+            val sentToESID = legacyAppConfig.sentToEntitySetId
+            val participantsESID = coreAppConfig.participantEntitySetId
+
+            val participantEKID =
+                enrollmentManager.getParticipantEntityKeyId(organizationId, messageOutcome.studyId, messageOutcome.participantId)
+
+            val messageEntity = mapOf(
+                edmCacheManager.getPropertyTypeId(EdmConstants.TYPE_FQN) to ImmutableSet.of<Any>(messageOutcome.messageType),
+                edmCacheManager.getPropertyTypeId(EdmConstants.OL_ID_FQN) to ImmutableSet.of<Any>(messageOutcome.sid),
+                edmCacheManager.getPropertyTypeId(EdmConstants.DATE_TIME_FQN) to ImmutableSet.of<Any>(messageOutcome.dateTime),
+                edmCacheManager.getPropertyTypeId(EdmConstants.DELIVERED_FQN) to ImmutableSet.of<Any>(messageOutcome.isSuccess)
```

Comment:
no change required. ImmutableSet.of<Any>(messageOutcome.messageType) could be replaced with setOf(messageOutcome)